### PR TITLE
Migrate `run-make/issue-30063` to `rmake`

### PR DESCRIPTION
**Pull request by Oneirical**
Part of #121876 and the associated [Google Summer of Code project](https://blog.rust-lang.org/2024/05/01/gsoc-2024-selected-projects.html).

(Sorry about the [inconvenience](https://github.com/rust-lang/rust/pull/125224#issuecomment-2118340932) of all these PRs, this is the last one batched for today. I will discuss how we can cut these down a bit.)

The last check was previously commented out in the Makefile, and I have readded it. If it fails the CI, this can be reconsidered.

### CODE REVIEW

File: tests/run-make/reset-codegen-1/rmake.rs [hunk 1]
```diff
@@ -0,0 +1,36 @@
+// When rustc receives 4 codegen-units, an output path and an emit flag all simultaneously,
+// this could cause an annoying recompilation issue, uselessly lengthening the build process.
+// This test tries to reproduce this issue with varied emit flags - if it passes, the codegen-units
+// are being appropriately reset to 1 after each pass.
```
Comment (jieyouxu) [inline tests/run-make/reset-codegen-1/rmake.rs:4]: AFAICT, this isn't actually being checked. We're just checking here that `foo.rs` "compiles with various cli flags", am I understanding that correctly?

Comment (jieyouxu) [reply tests/run-make/reset-codegen-1/rmake.rs:4]: As in, the original test also isn't checking what this comment is saying?

Comment (Oneirical) [reply tests/run-make/reset-codegen-1/rmake.rs]: You're right, I wrongfully assumed. The test is checking that the fix did not break compilation, not that the fix itself is doing what it is supposed to do (preventing the annoying re-compilation).

File: tests/run-make/reset-codegen-1/rmake.rs [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+// See https://github.com/rust-lang/rust/issues/30063
+
+//@ ignore-cross-compile
+
+use run_make_support::{rustc, tmp_dir};
+use std::fs;
+
+fn compile(emit: Option<&str>, output_dir: &str) {
+    let mut rustc = rustc();
+    let rustc = rustc.codegen_units(4).output(tmp_dir().join(output_dir)).input("foo.rs");
```
Comment (jieyouxu) [inline tests/run-make/reset-codegen-1/rmake.rs:15]: Question: is this `output_dir` a directory or a file?

File: tests/run-make/reset-codegen-1/rmake.rs [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        ("bc-output", Some("llvm-bc")),
+        ("ir-output", Some("llvm-ir")),
+        ("link-output", Some("link")),
+        ("obj-output", Some("obj")),
+        ("dep-output", Some("dep-info")),
+        ("multi-output", Some("asm,obj")),
+    ];
+    for (output_dir, emit) in flags {
+        fs::remove_file(output_dir).unwrap_or_default();
+        compile(emit, output_dir);
```
Comment (jieyouxu) [inline tests/run-make/reset-codegen-1/rmake.rs:35]: Nit: flipping `(output_dir, emit)` order between the for loop here and `compile()`'s argument order is a bit confusing

File: tests/run-make/reset-codegen-1/rmake.rs (hunk 2 above)
Comment (Oneirical) [reply tests/run-make/reset-codegen-1/rmake.rs]: You're right, it's a file, specified by the `-o` rustc flag and it's being removed individually by `remove_file`. Renamed.

File: tests/run-make/reset-codegen-1/rmake.rs (hunk 3 above)
Comment (Oneirical) [reply tests/run-make/reset-codegen-1/rmake.rs]: Swapped.

**Comment by Oneirical** [comment]
Revisions made.

@rustbot review

**Comment by jieyouxu** [comment]
Thanks, r=me after squashing commits into one.
@bors delegate+

**Pull request merged**